lmnet/lmnet/utils/tfds_builders/object_detection.py

```diff
@@ -0,0 +1,110 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+import math
+
+import tensorflow_datasets as tfds
+
+
+class ObjectDetectionBuilder(tfds.core.GeneratorBasedBuilder):
+    """
+    A custom TFDS builder for object detection dataset.
+    This class loads data from existing dataset classes and
+    generate TFDS formatted dataset which is equivalent to the original one.
+    See also: https://www.tensorflow.org/datasets/add_dataset
+    """
+
+    VERSION = tfds.core.Version("0.1.0")
+
+    def __init__(self, dataset_name, dataset_class=None, dataset_kwargs=None, **kwargs):
+        self.name = dataset_name
+        self.dataset_class = dataset_class
+        self.dataset_kwargs = dataset_kwargs
+        super().__init__(**kwargs)
+
+    def _info(self):
+        return tfds.core.DatasetInfo(
+            builder=self,
+            description="Custom TFDS dataset for object detection",
+            features=tfds.features.FeaturesDict({
+                "image": tfds.features.Image(),
+                "objects": tfds.features.SequenceDict({
+                    "label": tfds.features.ClassLabel(),
+                    "bbox": tfds.features.BBoxFeature(),
+                }),
+            }),
+        )
+
+    def _split_generators(self, dl_manager):
+        available_splits = {
+            "train": tfds.Split.TRAIN,
+            "validation": tfds.Split.VALIDATION,
+            "test": tfds.Split.TEST,
+        }
+
+        # Try to instantiate each subsets and skip the subset if it fails.
+        splits = []
+        for subset in self.dataset_class.available_subsets:
+            if subset in available_splits:
+                try:
+                    dataset = self.dataset_class(subset=subset, **self.dataset_kwargs)
+                except Exception:
+                    continue
+
+                self.info.features["objects"]["label"].names = dataset.classes
+
+                splits.append(
+                    tfds.core.SplitGenerator(
+                        name=available_splits[subset],
+                        num_shards=self._num_shards(dataset),
+                        gen_kwargs=dict(dataset=dataset)
+                    )
+                )
+
+        return splits
+
+    def _generate_examples(self, dataset):
+        for image, annotations in dataset:
+            height, width, _ = image.shape
+
+            objects = []
+            for annotation in annotations:
```

Comment:
Well, not sure if this is easier to read but i like
`objects = [{...} for xmin, ymin, w, h, label in annotation if label != -1]`